include/biscuit/literal.hpp

```diff
@@ -0,0 +1,168 @@
+     */
+    [[nodiscard]] bool IsResolved() const noexcept {
+        return m_offsets.empty();
+    }
+
+    /**
+     * Determines whether or not this literal is unresolved.
+     *
+     * A literal is considered unresolved if it still has any unhandled referencing offsets.
+     */
+    [[nodiscard]] bool IsUnresolved() const noexcept {
+        return !IsResolved();
+    }
+
+    /**
+     * Retrieves the location for this literal.
+     *
+     * @note If the returned location is empty, then this literal has not been assigned
+     *       a location yet.
+     */
+    [[nodiscard]] Location GetLocation() const noexcept {
+        return m_location;
+    }
+
+private:
+    // A literal instance is inherently bound to the assembler it's
+    // used with, as the offsets within the literal set depend on
+    // said assemblers code buffer.
+    friend class Assembler;
+
+    /**
+     * Places a literal to the given location.
+     *
+     * @param offset The offset to place this literal at.
+     *
+     * @returns The literal value so it can be copied to memory by the assembler.
+     *
+     * @pre The literal must not have already been placed at a previous location.
+     *      Attempting to place a literal multiple times is typically, in almost all scenarios,
+     *      the source of bugs.
+     *      Attempting to place an already placed literal will result in an assertion
+     *      being triggered.
+     */
+    [[nodiscard]] const T& Place(LocationOffset offset) noexcept {
+        BISCUIT_ASSERT(!IsPlaced());
+        m_location = offset;
+        return m_value;
+    }
+
+    /**
+     * Marks the given address as dependent on this literal.
+     *
+     * This is used in scenarios where a literal exists, but has not yet been
+     * placed at a location yet. It's important to track these addresses,
+     * as we'll need to patch the dependent load instructions with the
+     * proper offset once the literal is finally placed by the assembler.
+     *
+     * During literal placement, the offset will be calculated and inserted
+     * into dependent instructions.
+     */
+    void AddOffset(LocationOffset offset) {
+        // If a literal is already placed at a location, then offset tracking
+        // isn't necessary. Tripping this assert means we have a bug somewhere.
+        BISCUIT_ASSERT(!IsPlaced());
+        BISCUIT_ASSERT(IsNewOffset(offset));
+
+        m_offsets.insert(offset);
+    }
+
+    // Clears all the underlying offsets for this literal.
+    void ClearOffsets() noexcept {
+        m_offsets.clear();
+    }
+
+    // Determines whether or not this address has already been added before.
+    [[nodiscard]] bool IsNewOffset(LocationOffset offset) const noexcept {
+        return m_offsets.find(offset) == m_offsets.cend();
+    }
+
+    std::set<LocationOffset> m_offsets;
+    Location m_location;
+    const T m_value;
+};
+
+using Literal64 = Literal<uint64_t>;
```

Comment:
Should this not support 32-bit literals, etc as well? It's a little weird to have a generic templated type, but only provide an alias for one type of sized literal.

Reply:
I thought that there's no need for 32-bit literals, as those can be loaded into a register with 2 instructions anyway (LUI+ADDI).

It should support 128-bit+ literals though

Reply:
Might be a little weird for a user to try and use <= 32-bit literals and have it be a little wonky. You may want to add a static assert to forbid creating literals of those types at compile time (w/ a message explaining why they're not allowed) if you want to enforce that.

Reply:
Done. I should ask, what type do you think should be used for vector literals? Could be array, tuple, some custom struct, ...

Reply:
I would probably structure it like some variant of below, so that we can allow the user to structure their data however they like, without us needing to introduce a particular type that they absolutely need to use. It'd also allow easy support for arbitrary data sizes 64-bit >=

```cpp
// pretend we have large values here, {0, 1, 2, 3} just for brevity
// this could also be a struct or some other trivially copyable thing.
constexpr std::array<uint64_t, 4> big_data{0, 1, 2, 3};

void SomeFunc() {
  // Could probably have a template deduction guide to avoid needing users to specify type in <>
  Literal big_label(big_data);

  // ...

  Place(&big_label);
}
```

then in the assembler, we 'pull out' the generic ResolveLiteralOffsets like so:

```cpp
class Assembler {
public:
  // ... lot of code

private:
  template <typename T>
  void ResolveLiteralOffsets(Literal<T>* literal) {
    // Pass in the necessary internals here.
    ResolveLiteralRaw(&literal.m_value, sizeof(literal.m_value));
  }

  // actual implementation so we don't need to worry about template visibility for linking 
  // Add extra params as necessary here.
  void ResolveLiteralRaw(void* data, size_t data_size, ...);
};

Reply:
Essentially, you just want to have a way to pass-through the raw data itself and the necessary info outside of the templated literal type, since all that pragmatically needs to be done is the copying, and that's fairly type-agnostic to begin with.

Reply:
Cool, did some stuff akin to this. Also removed the Literal64 typedef, looks nicer now, thanks for the suggestions!